packages/react-dom/src/__tests__/DOMPropertyOperations-test.js

```diff
@@ -1235,6 +1268,95 @@ describe('DOMPropertyOperations', () => {
       expect(customelement.getAttribute('oncustomevent')).toBe(null);
     });
 
+    it('should only set complex values as properties', async () => {
+      const container = document.createElement('div');
+      const root = ReactDOMClient.createRoot(container);
+      const obj = {hello: 'world'};
+      const array = ['one', 'two'];
+      const date = new Date();
+
+      // First render before props are defined
+      await act(() => {
+        root.render(
+          <my-custom-element
+            obj={obj}
+            array={array}
+            date={date}
+          />,
+        );
+      });
+      const customElement = container.querySelector('my-custom-element');
+      expect(customElement.getAttribute('obj')).toBe(null);
+      expect(customElement.getAttribute('array')).toBe(null);
+      expect(customElement.getAttribute('date')).toBe(null);
+      // Install a setter to activate the `in` heuristic
+      Object.defineProperty(customElement, 'obj', {
+        set: function (x) {
+          this._obj = x;
+        },
+        get: function () {
+          return this._obj;
+        },
+      });
+      Object.defineProperty(customElement, 'array', {
+        set: function (x) {
+          this._array = x;
+        },
+        get: function () {
+          return this._array;
+        },
+      });
+      Object.defineProperty(customElement, 'date', {
+        set: function (x) {
+          this._date = x;
+        },
+        get: function () {
+          return this._date;
+        },
+      });
+
+      const obj2 = {hello: 'world'};
```

Comment:
NOTE: Needed new values for reconciliation purposes. This does make me a bit nervous per @josepharhar prior convo re: ensuring the custom element has been defined + upgraded by the time it is rendered. 😬. Still, this is an improvement overall, since that issue would occur regardless of changes in logic.